zeth-contracts/contracts/BaseMixer.sol

```diff
@@ -103,15 +127,134 @@ contract BaseMixer is MerkleTreeMiMC7, ERC223ReceivingContract {
+    // [Root, NullifierS, CommitmentS, h_sig, h_iS, Residual Field Element(S)]
     // ie, below is the index mapping of the primary input elements on the protoboard:
     // - Index of the "Root" field elements: {0}
     // - Index of the "NullifierS" field elements: [1, NumInputs + 1[
     // - Index of the "CommitmentS" field elements: [NumInputs + 1, NumOutputs + NumInputs + 1[
-    // - Index of the "v_pub_in" field element: {NumOutputs + NumInputs + 1}
-    // - Index of the "v_pub_out" field element: {NumOutputs + NumInputs + 1 + 1}
+    // - Index of the "h_sig" field element: {NumOutputs + NumInputs + 1}
+    // - Index of the "Message Authentication TagS" (h_i) field elements: [NumOutputs + NumInputs + 1 + 1, NumOutputs + NumInputs + 1 + NumOuputs [
+    // - Index of the "Residual Field Element(s)" field elements: [NumOutputs + NumInputs + 1 + NumOuputs + 1 , NumOutputs + NumInputs + 1 + NumOuputs + 1 + nb_field_residual]
+    //
+    // The Residual field elements are structured as follows:
+    // - v_pub_in [0, size_value[
+    // - v_pub_out [size_value + 1, 2*size_value[
+    // - h_sig remaining bits [2*size_value + 1, 2*size_value + (digest_length-field_capacity)[
+    // - nullifierS remaining bits [2*size_value + (digest_length-field_capacity) + 1, 2*size_value + (1+NumInputs)*(digest_length-field_capacity)[
+    // - commitmentS remaining bits [2*size_value + (1+NumInputs)*(digest_length-field_capacity) + 1, 2*size_value + (1+NumInputs+NumOutputs)*(digest_length-field_capacity)[
+    // - message authentication tagS remaining bits [2*size_value + (1+NumInputs+NumOutputs)*(digest_length-field_capacity) + 1, 2*size_value + (1+2*NumInputs+NumOutputs)*(digest_length-field_capacity)]
     // ============================================================================================ //
 
+    // This function is used to extract the public values (vpub_in and vpub_out) from the residual field element(S)
+    function assemble_public_values(uint[] memory primary_inputs) public view returns (uint64 vpub_in, uint64 vpub_out){
+        // We know vpub_in corresponds to the first 64 bits of the first residual field element after padding.
+        // We start by computing the padding .
+        uint padding = packing_residue_length;
+        // Update the padding if the public values and remainder bits fit in one field element
+        if (2*size_value+padding*(1+2*jsIn+jsOut) < field_capacity) {
+            padding = digest_length - (2*size_value + packing_residue_length*(1+2*jsIn+jsOut));
+        }
+
+        // We retrieve the public value in, remove any extra bits (due to the padding) and inverse the bit order
+        bytes32 vpub_bytes = (bytes32(primary_inputs[1 + 1 + 2*jsIn + jsOut]) << padding) >> (digest_length-size_value);
+        vpub_bytes = Bytes.swap_bit_order(vpub_bytes) >> (digest_length-size_value);
+        vpub_in = Bytes.get_int64_from_bytes8(Bytes.int256ToBytes8(uint(vpub_bytes)));
```

Comment:
The shift left and then back down seems to be unnecessary.  Can't you just do:
```
        bytes32 vpub_bytes = (bytes32(primary_inputs[1 + 1 + 2*jsIn + jsOut]) << padding);
        vpub_bytes = Bytes.swap_bit_order(vpub_bytes);
        vpub_in = Bytes.get_int64_from_bytes8(Bytes.int256ToBytes8(uint(vpub_bytes)));
```

Reply:
BTW, we can leave optimizations for #138, so no need to change this now if you find it clearer.

Reply:
No problem, this works well. I might just add a comment that `int256ToBytes8` extracts the last 8 bytes.

N.B. if we want to reverse the bit order, I guess we can just do:
`
        uint residual_hash_size = packing_residue_length*(1+2*jsIn+jsOut);
        bytes32 vpub_bytes = (bytes32(primary_inputs[1 + 1 + 2*jsIn + jsOut]) >> residual_hash_size + public_value_length);
        vpub_in = Bytes.get_int64_from_bytes8(Bytes.int256ToBytes8(uint(vpub_bytes)));

        vpub_bytes = bytes32(primary_inputs[1 + 1 + 2*jsIn + jsOut]) >> (residual_hash_size);
        vpub_out = Bytes.get_int64_from_bytes8(Bytes.int256ToBytes8(uint(vpub_bytes)));
`